### PR TITLE
SVHN dataset for torchvision

### PR DESCRIPTION
**Pull request by uridah**
In reference to https://github.com/pytorch/vision/issues/59
cc: @mjpieters

### CODE REVIEW

**Comment by soumith** [comment]
@uridah do you have a notebook that checks for sanity of this code, like the notebook that loads some CIFAR10 images and displays them:
https://github.com/pytorch/vision/blob/master/test/sanity_checks.ipynb

**Comment by uridah** [comment]
@soumith I don't but I can create one.

**Comment by mjpieters** [comment]
I don't know if PyTorch follows PEP8 as a styleguide? If so, you way want to run [flake8](http://flake8.pycqa.org/en/latest/) on your code and fix the issues that finds; I see a few whitespace issues in it. :-) If not, I'll leave it to the project owners to set the requirements.

**Comment by soumith** [comment]
thanks @mjpieters . We have a LINT check as part of the contbuild, and it is failing:

https://travis-ci.org/pytorch/vision/builds/209884377

Locally doing:
```
pip install flake8
flake8 .
``
will show exact LINT errors you need to fix @uridah

File: torchvision/datasets/svhn.py [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+            self.file_md5 = "e26dedcc434d2e4c54c9b2d4a06d8373"
+
+            if download:
+                self.download()
+
+            if not self._check_integrity():
+                raise RuntimeError('Dataset not found or corrupted.' +
+                               ' You can use download=True to download it')
+            
+            self.train_data = []
```
Comment (fmassa) [inline torchvision/datasets/svhn.py]: those two lines are not necessary, as they are assigned just below

File: torchvision/datasets/svhn.py [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+    file_md5 = ""
+    
+    def __init__(self, root, dataset='train', transform=None, target_transform=None, download=False):
+        self.root = root
+        self.transform = transform
+        self.target_transform = target_transform
+        self.dataset = dataset  # training set or test set or extra set
+        
+        # download and load the data 
+        if self.dataset=='train':
```
Comment (fmassa) [inline torchvision/datasets/svhn.py]: Maybe you could condensate these code-paths into a single one?
The only difference that I see is that you change the values of `url`, `filename` and `file_md5` for each set.
Maybe you could instead have a dictionary for each one of those fields, where the key is the set type (`train` / `extra`, `test`), and the values are the different urls, like done in [here](https://github.com/pytorch/vision/pull/63/files#diff-629efaeb461f617e4cdba4ef6b18222eR16)? Then you could have 
```python
self.url = url[dataset]
```

File: torchvision/datasets/svhn.py [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+
+        if self.target_transform is not None:
+            target = self.target_transform(target)
+
+        return img, target
+     
+
+    def __len__(self):
+        if self.dataset == 'train':
+            return 73257
```
Comment (fmassa) [inline torchvision/datasets/svhn.py]: maybe it's better to just return `len(self.train_data)`?

File: torchvision/datasets/svhn.py [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+class SVHN(data.Dataset):
+    url = ""
+    filename = ""
+    file_md5 = ""
+    
+    def __init__(self, root, dataset='train', transform=None, target_transform=None, download=False):
+        self.root = root
+        self.transform = transform
+        self.target_transform = target_transform
+        self.dataset = dataset  # training set or test set or extra set
```
Comment (fmassa) [inline torchvision/datasets/svhn.py]: maybe `dataset` is a misleading name, and something like `split` would be better?

**Comment by fmassa** [comment]
@uridah I made some small inline comments (very minor and are only style changes). The PR looks good, thanks!

**Comment by uridah** [comment]
@fmassa Your comments very very useful and really helped concise the code
@soumith I added a new notebook named sanity_checks1.ipynb which is basically same as the sanity_checks.ipynb notebook, but in addition, is also calling SVHN datasets for transformations https://github.com/uridah/vision/blob/master/test/sanity_checks1.ipynb
Also, ran the code through flake8
@mjpieters

File: torchvision/datasets/svhn.py [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+                self.download()
+
+            if not self._check_integrity():
+                    raise RuntimeError('Dataset not found or corrupted.' +
+                                       ' You can use download=True to download it')
+
+            # reading(loading) mat file as array
+            loaded_mat = sio.loadmat(os.path.join(root, self.filename))
+
+            if self.split != 'test':
```
Comment (fmassa) [inline torchvision/datasets/svhn.py]: I think we don't need to separate between `self.train_data` and `self.test_data`, and just `self.data` is fine.
This way you avoid this condition and the one in `__getitem__` as well.

File: torchvision/datasets/svhn.py [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+                self.train_data = loaded_mat['X']
+                self.train_labels = loaded_mat['y']
+                self.train_data = np.transpose(self.train_data, (3, 2, 1, 0))
+            else:
+                self.test_data = loaded_mat['X']
+                self.test_labels = loaded_mat['y']
+                self.test_data = np.transpose(self.test_data, (3, 2, 1, 0))
+        else:
+            print ("Wrong dataset entered! Please use split=train or split=extra or split=test")
```
Comment (fmassa) [inline torchvision/datasets/svhn.py]: It might be better to raise an error in here.
Maybe something like
```python
raise ValueError("Wrong dataset entered! Please use split=train, split=extra or split=test")
```

File: torchvision/datasets/svhn.py [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+        'extra': ["http://ufldl.stanford.edu/housenumbers/extra_32x32.mat",
+                  "extra_32x32.mat", "a93ce644f1a588dc4d68dda5feec44a7"]}
+
+    def __init__(self, root, split='train', transform=None, target_transform=None, download=False):
+        self.root = root
+        self.transform = transform
+        self.target_transform = target_transform
+        self.split = split  # training set or test set or extra set
+
+        if self.split in self.split_list:
```
Comment (fmassa) [inline torchvision/datasets/svhn.py]: I think it might be better to have a condition in the other way around. Something like
```python
if self.split not in self.split_list:
    raise ValueError("...")

# rest of the code goes here, without the extra identation
```

File: torchvision/datasets/svhn.py [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+                self.train_data = np.transpose(self.train_data, (3, 2, 1, 0))
+            else:
+                self.test_data = loaded_mat['X']
+                self.test_labels = loaded_mat['y']
+                self.test_data = np.transpose(self.test_data, (3, 2, 1, 0))
+        else:
+            print ("Wrong dataset entered! Please use split=train or split=extra or split=test")
+
+    def __getitem__(self, index):
+        if self.split == 'train' or self.split == 'extra':
```
Comment (fmassa) [inline torchvision/datasets/svhn.py]: If you remove the difference between `self.train_data` and `self.train_labels` and add only `self.data` and `self.labels`, this condition can be removed and the code is a bit simpler

File: torchvision/datasets/svhn.py [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+        if self.transform is not None:
+            img = self.transform(img)
+
+        if self.target_transform is not None:
+            target = self.target_transform(target)
+
+        return img, target
+
+    def __len__(self):
+        return len(self.train_data)
```
Comment (fmassa) [inline torchvision/datasets/svhn.py]: Another reason why it might be better to merge `train_data` into `data` only. This currently doesn't work in the `test` case.

**Comment by fmassa** [comment]
Hi @uridah 
I made some more comments. The code is looking very nice!

**Comment by uridah** [comment]
Thanks @fmassa. Updated according to your suggestions

File: torchvision/datasets/svhn.py [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+
+    def __init__(self, root, split='train', transform=None, target_transform=None, download=False):
+        self.root = root
+        self.transform = transform
+        self.target_transform = target_transform
+        self.split = split  # training set or test set or extra set
+
+        if self.split not in self.split_list:
+            raise ValueError('Wrong split entered! Please use split=train or split=extra or split=test')
+        else:
```
Comment (fmassa) [inline torchvision/datasets/svhn.py]: One last thing, this `else` (and the idented code) is not needed anymore, because of the `raise` in the line just above.

File: torchvision/datasets/svhn.py [hunk 11]
```diff
@@ -0,0 +1,109 @@
...
+
+        self.url = self.split_list[split][0]
+        self.filename = self.split_list[split][1]
+        self.file_md5 = self.split_list[split][2]
+
+        if download:
+            self.download()
+
+        if not self._check_integrity():
+                raise RuntimeError('Dataset not found or corrupted.' +
```
Comment (fmassa) [inline torchvision/datasets/svhn.py]: not important: the indentation in here is off

**Comment by fmassa** [comment]
About the `check_sanity1.pynb` file, maybe you could integrate your changes into the `check_sanity.pynb` file? What do you think?

**Comment by uridah** [comment]
I will, if you guys think it's good enough.

**Comment by fmassa** [comment]
@uridah another thing, it would be great if you could add an entry in the doc in `README.md`, like the one in `Cifar10` for example. Maybe an example in the doc would be enough?
cc @soumith

**Comment by uridah** [comment]
here is the pull request for the change in documentation: https://github.com/pytorch/vision/pull/104
@fmassa

**Comment by fmassa** [comment]
Hi @uridah 
So, looking at the notebook that you added, I have the impression that the numbers are rotated. Maybe the `x` and `y` axis in the dataset are inverted, and you need to transpose them?
Also, could you please fix the indentation in the line I commented?
Once those are fixed, I think there is no need for the notebook (but @soumith knows better), so could you please remove it?
After that, the PR is good to be merged!

**Comment by uridah** [comment]
@fmassa as it turns out I needed it to transpose along 3,2,0,1 axis instead of 3,2,1,0. I have fixed that and updated the indentation and sanity_checks1.ipynb. Please have a look and let me know if anything else needs to be changed.
Also, I created a separate request to update the documentation but now I have merged it with this one.

**Comment by soumith** [comment]
Thanks Uridah, as you saw the last 4 commits, I made some minor changes to your PR. But it looked great. Merged into master now!!!

**Comment by vabh** [comment]
Hello,

It seems that the labels returned are in the range `1-10`  (the data set assigns class `10` to the digit `0` and class `d` for all other digits, `d` [ http://ufldl.stanford.edu/housenumbers/ see section `overview` ]).

Given that some of the loss functions (CELoss, NLLLoss) expect the class labels to be in the range `[0, C-1]`, shouldn't that be reflected here as well?

Another thing is that the the returned `labels` are of type ByteTensor and of size `batchSize x 1`. Again, CELoss, etc, expect it to be a 1d `LongTensor`.

I can make a PR if the current behaviour is inconsistent and should be made similar to the other data sets, for example as in CIFAR10

**Comment by soumith** [comment]
@vabh if you could make a PR to reflect this, that'd be great. Thanks.

**Comment by vabh** [comment]
@soumith I made one here: #194
